### PR TITLE
Add --depfile option

### LINKED ISSUE #192: Add support for outputting dependencies
It would be nice if cbindgen could provide the full list of input files it used to generate proper dependencies for build systems. @emilio [is adding cbindgen to the Firefox build](https://phabricator.services.mozilla.com/D2447) and his glue code is manually providing dependencies which makes me worry that that list could be incorrect, which would result in broken builds.

### LINKED ISSUE #815: Add optional depfile output
Hi,

I'm working on integrating cbindgen into a foreign build system (CMake), and it would be great if `cbindgen` could output a depfile listing the input source files, so my build system knows when to rerun cbindgen for a given crate.
I would also be willing to work on a PR for this feature, if the chances are good for this feature request to be accepted.

### PR DESCRIPTION
**Pull request by jschwe**
Add an option to output a depfile for outside build-systems to learn the file dependencies of the bindings.

cbindgen already keeps track of all the source-files it visits, the information just had to be collected and passed up the callchain.

The output depfile is the common Makefile `.d` file, which cargo also already produes. example:
```
/path/to/generated_header.h: /path/to/src1 /path/to/src2
```

Whitespace in filepaths is escaped to support filepaths containing whitespace. This requires assuming an encoding, so we assume the filepath is utf-8 compliant (and otherwise lossy convert to utf-8).

Since the left hand side requires a file, it only makes sense to generate a depfile if `--out` is also specified, so the current implementation just ignores `--depfile` if `--out` is missing.

I looked at the sorrounding code, and cbindgen is generally very liberal with `unwrap`, so I also didn't add any error handling for this PR. If the filepaths contain non-unicode characters, I chose to use the `lossy` variant for conversion,

I tested this with CMake and Ninja on Linux, and the integration seems to be working as expected (also with paths containing spaces).
I'm not sure yet how to best add tests to cbindgen itself though - the tests seem to be generated from the build-script and I'm not quite sure how to add my one for this feature. Any pointers regarding would be appreciated.

Closes #815 
Closes #192

### CODE REVIEW

File: src/bindgen/bindings.rs
Comment (emilio) [inline src/bindgen/bindings.rs:178]: Doesn't closing the file flush as well? I doubt this is needed right?

File: src/bindgen/parser.rs
Comment (emilio) [inline src/bindgen/parser.rs:184]: So this order is basically random, I'd rather have a well defined order by pushing the path in the `Entry::Vacant` case in `parse_mod`. Otherwise let's have some defined order in the depfile generation?

**Review by emilio — CHANGES_REQUESTED** [review]
Thanks, looks reasonable, but I think we should have a defined order to avoid the ordering depending on cache keys / hashmap iteration order.

It'd also be great to have a test for this.

File: src/bindgen/bindings.rs
Comment (jschwe) [reply src/bindgen/bindings.rs:178]: When the file is dropped, errors when closing are ignored. Manually flushing and unwrapping results in a panic for errors. I don't have a strong opinion here, since errors when closing the file should be extremely rare. I can remove this line if you prefer.

File: src/bindgen/parser.rs
Comment (jschwe) [reply src/bindgen/parser.rs:184]: I pushed a commit which sorts the output before writing. This way all source files are sorted, and the config file (if any) is the last entry. This should make writing tests much easier, since it allows us to just compare an output file with an expected file.

**Comment by jschwe** [comment]
> It'd also be great to have a test for this.

Do you have any recommendations for a test? Should I create a new integration test in tests, or integrate this into the `tests` test?

**Comment by jschwe** [comment]
Update: This PR currently doesn't support expanding (since we use `rustc` to expand the crate, and we don't know the source files ourselves). Probably this can be solved by letting rustc itself emit a `.d` file during the expansion and parsing that too.

Maybe it would be better to put that into a seperate PR though, to keep the diff of this one smaller?

**Comment by jschwe** [comment]
I added some tests which use CMake to verify that the generated depfile works as expected (i.e. the bindings aren't regenerated on a second build).
I previously also added `--depfile` to the tests in `tests` in commit [test --depfile](https://github.com/eqrion/cbindgen/pull/820/commits/18351e0f71e1cf5d9df0528bd5a7b1452e7e5e8b), but I'm leaning towards dropping that commit and only keeping the new CMake based test for the --depfile option. What do you think?

I also discovered that implicit configs (`cbindgen.toml` in the root dir) where previously not put into the depfile, so I added a commit which addresses that. Since the `Config` struct and methods are public, I added new internal methods to avoid any semver breaking changes.

**Comment by jschwe** [comment]
friendly ping

**Comment by jschwe** [comment]
@emilio Could I get another review round? Changes since your last review are summarized in this comment: https://github.com/eqrion/cbindgen/pull/820#issuecomment-1465158996

**Comment by jschwe** [comment]
@emilio Could I get another review round? Changes since your last review are summarized in this comment: https://github.com/mozilla/cbindgen/pull/820#issuecomment-1465158996

File: src/bindgen/parser.rs
Comment (emilio) [inline src/bindgen/parser.rs:83]: This order is not deterministic right? We should sort these or something.

File: src/bindgen/builder.rs
Comment (emilio) [inline src/bindgen/builder.rs:304]: Any reason we could not just store the `src_path: Option<PathBuf>` in the `Config`? That seems like it'd be cleaner?

**Review by emilio — CHANGES_REQUESTED** [review]
Sorry for the lag in re-reviewing this, personal life got in the way.

I think there are only a couple relatively minor issues.

I think making the path a (crate-private) field in the `Config` struct would simplify a bunch of the core failures.

Thanks for all the testing effort, much appreciated.

File: src/bindgen/builder.rs
Comment (jschwe) [reply src/bindgen/builder.rs:304]: The `config_src` should contain the path to the config file (if it exists). 
The `Config` itself is parsed from the config file. It would be strange if the user could set a path in the config file, which should point to the config file itself.

Comment (jschwe) [reply src/bindgen/builder.rs]: Finished the change, this is indeed much cleaner.

**Comment by jschwe** [comment]
Probably https://github.com/mozilla/cbindgen/issues/836 should be fixed first before merging, so that the CI can run again.

File: src/bindgen/parser.rs
Comment (emilio) [reply src/bindgen/parser.rs:83]: Ah, great, sorry I missed that!

File: src/main.rs
Comment (emilio) [inline src/main.rs:90]: This change looks unrelated now. Can we undo this? Changing the default behavior of config files is a bit subtle, and there's no test for this change.

File: tests/depfile.rs
Comment (emilio) [inline tests/depfile.rs:7]: `name` seems unused.

File: tests/tests.rs
Comment (emilio) [inline tests/tests.rs:12]: Oh, sweet!

File: src/main.rs
Comment (emilio) [reply src/main.rs:90]: (That said I think it's probably an uncontroversial improvement, I just rather let it land with a separate PR)

**Review by emilio — CHANGES_REQUESTED** [review]
Just one minor tweak. Let me look at fixing the actions. Looks good with those, thanks!

File: tests/depfile.rs
Comment (jschwe) [reply tests/depfile.rs]: Removed name

File: src/main.rs
Comment (jschwe) [reply src/main.rs]: I reverted it. I can open a separate MR afterwards.

**Comment by emilio** [comment]
Can you rebase? We should have GH actions now.

**Comment by jschwe** [comment]
Rebased and squashed into two commits.

**Review by emilio — APPROVED** [review]
Awesome, ty!

**Comment by wongchaiyatham-max** [comment]
> Add an option to output a depfile for outside build-systems to learn the file dependencies of the bindings.
> 
> cbindgen already keeps track of all the source-files it visits, the information just had to be collected and passed up the callchain.
> 
> The output depfile is the common Makefile `.d` file, which cargo also already produes. example:
> ```
> /path/to/generated_header.h: /path/to/src1 /path/to/src2
> ```
> 
> Whitespace in filepaths is escaped to support filepaths containing whitespace. This requires assuming an encoding, so we assume the filepath is utf-8 compliant (and otherwise lossy convert to utf-8).
> 
> Since the left hand side requires a file, it only makes sense to generate a depfile if `--out` is also specified, so the current implementation just ignores `--depfile` if `--out` is missing.
> 
> I looked at the sorrounding code, and cbindgen is generally very liberal with `unwrap`, so I also didn't add any error handling for this PR. If the filepaths contain non-unicode characters, I chose to use the `lossy` variant for conversion,
> 
> 
> I tested this with CMake and Ninja on Linux, and the integration seems to be working as expected (also with paths containing spaces).
> I'm not sure yet how to best add tests to cbindgen itself though - the tests seem to be generated from the build-script and I'm not quite sure how to add my one for this feature. Any pointers regarding would be appreciated.
> Invalids"roots-only™"6f2ddc538acb19180f4c8e96cff94ccf27822e8b
> Closes #815 
> Closes #192
>

**Pull request merged**
